File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        def dfs(node):
            if not node: return 0

            return max(dfs(node.left), dfs(node.right)) + 1
        
        if not root: return 0
        depth_current_node = dfs(root.left) + dfs(root.right)
        depth_left_subtree = self.diameterOfBinaryTree(root.left)
        depth_right_subtree = self.diameterOfBinaryTree(root.right)
        depth_current_subtree = max(depth_left_subtree, depth_right_subtree)

        return max(depth_current_node, depth_current_subtree)

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        List<Integer> pre = new ArrayList<>();
        for (int num: preorder) {
            pre.add(num);
        }
        return build(pre, inorder);
    }

    private TreeNode build(List<Integer> preorder, int[] inorder) {
        if (inorder.length == 0) return null;
        int rootVal = preorder.remove(0);
        int idx = 0;
        TreeNode root = new TreeNode(rootVal);
        while (rootVal != inorder[idx]) {
            idx++;
        }
        root.left = build(preorder, Arrays.copyOfRange(inorder, 0, idx));
        root.right = build(preorder, Arrays.copyOfRange(inorder, idx+1, inorder.length));
        return root;
    }
}
